### ISSUE TITLE
Oauth2 login return refreshToken: undefined

### ISSUE THREAD

**Issue by paustria** [jsforce/jsforce#702]
I have checked other issues and I don't see anything related to the way I called SF.

Here is the snippet that I wrote the function to login to SF.
```
const sfLogin = () => new Promise((resolve, reject) => {
  let conn = new jsforce.Connection({
    oauth2: {
      loginUrl: process.env.SALESFORCE_API_URL,
      clientId: process.env.SALESFORCE_CLIENT_ID,
      clientSecret: process.env.SALESFORCE_CLIENT_SECRET,
      redirectUri: 'https://localhost:8000/vendor/salesforce'
    }
  });

  const username = process.env.SALESFORCE_USERNAME;
  const password = process.env.SALESFORCE_PASSWORD;

  console.log('connect via OAuth2 Resource Owner');
  conn.login(username, password, (err, userInfo) => {
    if (err) {
      console.log('Salesforce connection issue.');

      reject(false);
    }

    console.log('conn2', conn);

    resolve(conn);
  });
});
```

on `console.log('conn2', conn);`

I got 
```
conn2 EventEmitter {
  _logger: { _logLevel: 2 },
  oauth2:
   { loginUrl: 'https://test.salesforce.com',
     authzServiceUrl: 'https://test.salesforce.com/services/oauth2/authorize',
     tokenServiceUrl: 'https://test.salesforce.com/services/oauth2/token',
     revokeServiceUrl: 'https://test.salesforce.com/services/oauth2/revoke',
     clientId: 'SOMECLIENTID',
     clientSecret: 'SOMECLIENTSECRET',
     redirectUri: undefined,
     _transport: {} },
  loginUrl: 'https://test.salesforce.com',
  version: '39.0',
  maxRequest: 10,
  _transport: {},
  callOptions: undefined,
  analytics: Analytics { _conn: [Circular] },
  apex: Apex { _conn: [Circular] },
  bulk: Bulk { _conn: [Circular], _logger: { _logLevel: 2 } },
  chatter: { _conn: [Circular] },
  metadata: { _conn: [Circular] },
  soap: { _conn: [Circular] },
  streaming: Streaming { _conn: [Circular] },
  tooling:
   Tooling {
     _conn: [Circular],
     _logger: { _logLevel: 2 },
     query: [Function],
     queryMore: [Function],
     create: [Function],
     insert: [Function],
     retrieve: [Function],
     update: [Function],
     upsert: [Function],
     del: [Function],
     delete: [Function],
     destroy: [Function],
     describe: [Function],
     describeGlobal: [Function],
     sobject: [Function],
     cache: Cache { _entries: [Object] },
     'describe$': { [Function: $fn] clear: [Function] },
     'describeSObject$': { [Function: $fn] clear: [Function] },
     describeSObject: [Function],
     'describeGlobal$': { [Function: $fn] clear: [Function] },
     sobjects: {} },
  process:
   { rule: ProcessRule { _conn: [Circular] },
     approval: ApprovalProcess { _conn: [Circular] } },
  cache: Cache { _entries: { describeGlobal: [Object] } },
  _refreshDelegate: SessionRefreshDelegate { _conn: [Circular], _refreshFn: [Function], _refreshing: false },
  'describe$': { [Function: $fn] clear: [Function] },
  describe: [Function],
  'describeSObject$': { [Function: $fn] clear: [Function] },
  describeSObject: [Function],
  'describeGlobal$': { [Function: $fn] clear: [Function] },
  describeGlobal: [Function],
  instanceUrl: 'https://xxx.my.salesforce.com',
  accessToken: '00D2C0000000uoO!AREAQCkfbYKUs_aji424ABwHJIaJlUka2oliNLVBn45czQbNVMISlKKuitfSySnqRYTNBVBsLy8wSI6bs2Rpv1.CU9Cc5m7N',
  refreshToken: undefined,
  signedRequest: undefined,
  limitInfo: {},
  sobjects: {},
  _sessionType: 'oauth2',
  userInfo:
   { id: 'SOMEID',
     organizationId: 'SOMEORGID',
     url: 'https://test.salesforce.com/id/SOMEID/SOMEORGID' } }
```

Here are `Selected OAuth Scopes` in SF
```
Access and manage your data (api)
Full access (full)
Perform requests on your behalf at any time (refresh_token, offline_access)
```

Are there any other steps to acquire `refreshToken`?

**Comment by paustint** [comment]
As noted in the docs here https://jsforce.github.io/document/#access-token-with-refresh-token
> NOTE: Refresh token is only available for OAuth2 authorization code flow.

When you use `conn.login()`, this is not the authorization flow, this is the `OAuth2 Resource Owner Password Credential` flow.  Since you have the username and password, then you don't really need a refresh token since you already have credentials to access the system - and if you are not storing the user credentials, then you should be using the OAuth authorization redirect flow so you can get the refresh token for offline access.

**Comment by paustria** [comment]
@paustint Does this mean I can keep calling login as much as I want? My worry is I create SalesForce Account on web account registration page. What if there are 1000 users registering concurrently within 10 seconds. Will Salesforce accept 1000 logins api within those timeframe?

**Comment by paustint** [comment]
Salesforce only rate limits API calls longer than 20 seconds, so you should
be ok there. You will be consuming extra API calls with the additional
logins.

Maybe you can cache the connection on the server and get a new access token
only when you get a 401 or you can keep track and call login at some
interval.

For reference, here are the relevant sfdc docs:
https://help.salesforce.com/articleView?id=remoteaccess_oauth_username_password_flow.htm&type=0

**Comment by paustria** [comment]
@paustint Thank you! I will probably do the interval route. I will close this issue now.

**Comment by Caperious** [comment]
I'm sorry for interrupting a closed issue, but I think I have a kinda similar problem.. 

I only have only one user on Salesforce, for whom we are pushing updates from an external API..
Currently, we are using `OAuth2 Resource Owner Password Credential flow`, but the problem is when the User updates the password without us knowing ( he always forgets to tell us.. ), the API stops updating SF. 

So basically we would need to get the refresh token, but don't want any user interaction or input to get it.. 

Is that even possible?

I'm sorry if this question is out of line or inappropriate, ill remove it, if that's the case
